e2e-tests/verify-is-synced/src/main.rs

```diff
@@ -0,0 +1,262 @@
+use bitcoin::{
+    blockdata::constants::genesis_block, consensus::Encodable, Address, Block, BlockHeader,
+    Network as BitcoinNetwork, OutPoint,
+};
+use candid::CandidType;
+use ic_btc_test_utils::{BlockBuilder, TransactionBuilder};
+use ic_cdk_macros::{init, update};
+use serde::{Deserialize, Serialize};
+use std::cell::{Cell, RefCell};
+use std::str::FromStr;
+
+type BlockBlob = Vec<u8>;
+type BlockHeaderBlob = Vec<u8>;
+type BlockHash = Vec<u8>;
+
+const ADDRESS_1: &str = "bcrt1qg4cvn305es3k8j69x06t9hf4v5yx4mxdaeazl8";
+const ADDRESS_2: &str = "bcrt1qxp8ercrmfxlu0s543najcj6fe6267j97tv7rgf";
+const ADDRESS_3: &str = "bcrt1qp045tvzkxx0292645rxem9eryc7jpwsk3dy60h";
+const ADDRESS_4: &str = "bcrt1qjft8fhexv4znxu22hed7gxtpy2wazjn0x079mn";
+const ADDRESS_5: &str = "bcrt1qenhfslne5vdqld0djs0h0tfw225tkkzzc60exh";
+
+#[derive(CandidType, Clone, Copy, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+enum Network {
+    #[serde(rename = "mainnet")]
+    Mainnet,
+    #[serde(rename = "testnet")]
+    Testnet,
+    #[serde(rename = "regtest")]
+    Regtest,
+}
+
+#[derive(CandidType, Clone, Debug, PartialEq, Eq, Deserialize)]
+enum GetSuccessorsRequest {
+    #[serde(rename = "initial")]
+    Initial(GetSuccessorsRequestInitial),
+    #[serde(rename = "follow_up")]
+    FollowUp(u8),
+}
+
+#[derive(CandidType, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+struct GetSuccessorsRequestInitial {
+    pub network: Network,
+    pub processed_block_hashes: Vec<BlockHash>,
+}
+
+#[derive(CandidType, Clone, Debug, Deserialize, Hash, PartialEq, Eq, Serialize)]
+enum GetSuccessorsResponse {
+    #[serde(rename = "complete")]
+    Complete(GetSuccessorsCompleteResponse),
+    #[serde(rename = "partial")]
+    Partial(GetSuccessorsPartialResponse),
+    #[serde(rename = "follow_up")]
+    FollowUp(BlockBlob),
+}
+
+#[derive(CandidType, Clone, Debug, Default, Deserialize, Hash, PartialEq, Eq, Serialize)]
+struct GetSuccessorsCompleteResponse {
+    blocks: Vec<BlockBlob>,
+    next: Vec<BlockHeaderBlob>,
+}
+
+#[derive(CandidType, Clone, Debug, Default, Deserialize, Hash, PartialEq, Eq, Serialize)]
+struct GetSuccessorsPartialResponse {
+    partial_block: BlockBlob,
+    next: Vec<BlockHeaderBlob>,
+    remaining_follow_ups: u8,
+}
+
+thread_local! {
+    static BLOCKS: RefCell<Vec<BlockBlob>> = RefCell::new(Vec::new());
+
+    static COUNT: Cell<u64> = Cell::new(0);
+
+    static BLOCK_HEADERS: RefCell<Vec<BlockHeaderBlob>> = RefCell::new(Vec::new());
+}
+
+// Initialize the blocks.
+#[init]
+fn init() {
```

Comment:
Do we need this relatively complicated logic? Why not just create empty blocks without any transactions?

Reply:
If I tried to create block without transactions I get following error:
`Creating UI canister on the local network.
The UI canister on the "local" network is "rrkah-fqaaa-aaaaa-aaaaq-cai"
Installing code for canister e2e-verify-is-synced, with canister ID rwlgt-iiaaa-aaaaa-aaaaa-cai
[Canister rwlgt-iiaaa-aaaaa-aaaaa-cai] Panicked at 'could not initialize thread_rng: All entropy sources failed (permanently unavailable); cause: OsRng: support for wasm32 requires emscripten, stdweb or wasm-bindgen (permanently unavailable)', /home/dduric/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.6.5/src/rngs/thread.rs:80:17
Error: Failed while trying to deploy canisters.
Caused by: Failed while trying to deploy canisters.
  Failed while trying to install all canisters.
    Failed to install wasm module to canister 'e2e-verify-is-synced'.
      Failed to install wasm in canister 'rwlgt-iiaaa-aaaaa-aaaaa-cai'.
        Failed to install wasm.
          The Replica returned an error: code 5, message: "Canister rwlgt-iiaaa-aaaaa-aaaaa-cai trapped explicitly: Panicked at 'could not initialize thread_rng: All entropy sources failed (permanently unavailable); cause: OsRng: support for wasm32 requires emscripten, stdweb or wasm-bindgen (permanently unavailable)', /home/dduric/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.6.5/src/rngs/thread.rs:80:17"`

Reply:
But if I put just one transaction per block it works, hence I did that.